plugins/spi/SPIOutput.cpp

```diff
@@ -486,6 +501,121 @@ uint8_t SPIOutput::P9813CreateFlag(uint8_t red, uint8_t green, uint8_t blue) {
+  // https://cpldcpu.wordpress.com/2014/11/30/understanding-the-apa102-superled/
+  // Data-Struct
+  // StartFrame: 4 bytes = 32 bits zeros (APA102_START_FRAME_BYTES)
+  // LEDFrame: 1 byte FF ; 3 bytes color info (Blue, Green, Red)
+  // EndFrame: (n/2)bits; n = pixel_count
+
+  // calculate DMX-start-address
+  const unsigned int first_slot = m_start_address - 1;  // 0 offset
+
+  // only do something if at least 1 pixel can be updated..
+  if (buffer.Size() - first_slot < APA102_SLOTS_PER_PIXEL) {
+    OLA_INFO << "Insufficient DMX data, required " << APA102_SLOTS_PER_PIXEL
+             << ", got " << buffer.Size() - first_slot;
+    return;
+  }
+
+  // We always check out the entire string length, even if we only have data
+  // for part of it
+  const unsigned int output_length = APA102_START_FRAME_BYTES +
+                                (m_pixel_count * APA102_SPI_BYTES_PER_PIXEL);
+  uint8_t *output = m_backend->Checkout(m_output_number, output_length,
+                                      CalculateAPA102LatchBytes(m_pixel_count));
+  // only update SPI data if possible
+  if (!output) {
+    return;
+  }
+
+  for (unsigned int i = 0; i < m_pixel_count; i++) {
+    // Convert RGB to APA102 Pixel
+    unsigned int offset = first_slot + (i * APA102_SLOTS_PER_PIXEL);
+
+    // We need to avoid the first 4 bytes of the buffer since that acts as a
+    // start of frame delimiter
+    unsigned int spi_offset = APA102_START_FRAME_BYTES +
+                              (i * APA102_SPI_BYTES_PER_PIXEL);
+    // set pixel data
+    // first Byte contains:
+    // 3 bits start mark (111) + 5 bits global brightness
+    // set global brightness fixed to 31 --> that reduces flickering
+    // that can be written as 0xE0 & 0x1F
+    output[spi_offset + 0] = 0xFF;
+    // only write pixel data if buffer has complete data for this pixel:
+    if ((buffer.Size() - offset) >= APA102_SLOTS_PER_PIXEL) {
+      // Convert RGB to APA102 Pixel
+      // skip spi_offset + 0 (is already set)
+      output[spi_offset + 1] = buffer.Get(offset + 2);  // blue
+      output[spi_offset + 2] = buffer.Get(offset + 1);  // green
+      output[spi_offset + 3] = buffer.Get(offset + 0);  // red
+    }
+  }
+  m_backend->Commit(m_output_number);
+}
+
+void SPIOutput::CombinedAPA102Control(const DmxBuffer &buffer) {
+  // for Protocol details see IndividualAPA102Control
+
+  const uint16_t first_slot = m_start_address - 1;  // 0 offset
+
+  // check if enough data is there.
+  if (buffer.Size() - first_slot < APA102_SLOTS_PER_PIXEL) {
+    OLA_INFO << "Insufficient DMX data, required " << APA102_SLOTS_PER_PIXEL
+             << ", got " << buffer.Size() - first_slot;
+    return;
+  }
+
+  // get data for entire string length
+  const uint16_t output_length = (m_pixel_count + 1) *
+      APA102_SPI_BYTES_PER_PIXEL;
+  uint8_t *output = m_backend->Checkout(m_output_number, output_length,
+      CalculateAPA102LatchBytes(m_pixel_count));
```

Comment:
sorry - i changed this after the comment from peternewman
https://github.com/OpenLightingProject/ola/pull/722#discussion-diff-31242838
how should i line it up correctly? (its to long for one line)